### ISSUE TITLE
Neighbourhood density suggestion

### ISSUE THREAD

**Issue by kchall** [PhonologicalCorpusTools/CorpusTools#404]
Suggestions from a user (that I know personally), with respect to the Korean corpus:

In any case, I was surprised that you said it took three days to calculate the neighbors. We have calculated neighbors on an old version of the corpus and it took at most a few hours, so I was trying to figure out where the slowdown came from. I took the liberty of looking through the neighborhood_density.py file in corpustools/neighdens/ on your git repo.

My understanding of your current calculation algorithm is as follows (simplifying some details, without loss of generality (I think)):
 - compute pairwise edit distance between every word in the corpus. For a corpus of N words, this requires (N^2-N)/2 comparisons (although if the pairwise comparisons aren't optimized it could be as many as N^2).
 - for each word, count how many other words have an edit distance at or below the threshold (conventionally 1).
 - this count is the density, and you can also return the actual neighbors themselves.

While this approach is the intuitive one (for me, at least), it requires many edit distance calculations, which are computationally taxing, and the number of calculations grows exponentially with the corpus size. An alternative algorithm, which is the one that we used earlier, appears to be somewhat faster:
 - for each word, compute all possible neighbors for the given edit distance threshold. E.g. for "cat" we compute every possible string with an edit distance of 1, given an alphabet of symbols, yielding "aat", "bat", "dat", ..., "cot", "cpt", "cqt", ..., "cax", "cay", "caz".
 - for each possible neighbor, look up in the lexicon if that neighbor actually exists. If so, we record it as a neighbor (and increment the neighbor count by 1), and if not, we discard.
This method requires more RAM, but releases the computer of having to make explicit edit distance calculations. Using this method we managed to calculate neighbors for an older version of the Korean corpus (a smaller version, admittedly) in just a few hours.

Anyway, if you are considering refactoring parts of the PCT, I would suggest you consider the pros and cons of these approaches. I'm not 100% sure that the method I outlined is faster, nor am I sure of how computation time increases with corpus size.

**Comment by mmcauliffe** [comment]
Yeah, all the previous version of neighborhood density calculations, I've used the second algorithm, and it is indeed a lot faster, especially when searching in a SQL database.  I think, from my rudimentary knowledge of algorithms, that the current algorithm is O(n^2) and the other algorithm is O(kn) where n is the number of words in the corpus and k is the number of segments in the inventory.  So the second algorithm would scale much better with larger corpora.
